File: bootstrap/src/main/java/net/dontdrinkandroot/wicket/bootstrap/component/button/AjaxButton.java
/*
 * Copyright (C) 2012-2017 Philip Washington Sorst <philip@sorst.net>
 * and individual contributors as indicated
 * by the @authors tag.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.dontdrinkandroot.wicket.bootstrap.component.button;

import net.dontdrinkandroot.wicket.bootstrap.behavior.ButtonBehavior;
import net.dontdrinkandroot.wicket.bootstrap.css.ButtonSize;
import net.dontdrinkandroot.wicket.bootstrap.css.ButtonStyle;
import org.apache.wicket.ajax.markup.html.AjaxLink;
import org.apache.wicket.model.IModel;

/**
 * @param <T> Type of the model object.
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public abstract class AjaxButton<T> extends AjaxLink<T> implements IButton
{
    protected ButtonBehavior buttonBehavior = new ButtonBehavior();

    public AjaxButton(String id)
    {
        super(id);
    }

    public AjaxButton(String id, IModel<T> model)
    {
        super(id, model);
    }

    @Override
    protected void onInitialize()
    {
        super.onInitialize();
        this.add(this.buttonBehavior);
    }

    @Override
    public ButtonSize getButtonSize()
    {
        return this.buttonBehavior.getButtonSize();
    }

    @Override
    public AjaxButton<T> setButtonSize(ButtonSize buttonSize)
    {
        this.buttonBehavior.setButtonSize(buttonSize);
        return this;
    }

    @Override
    public ButtonStyle getButtonStyle()
    {
        return this.buttonBehavior.getButtonStyle();
    }

    @Override
    public AjaxButton<T> setButtonStyle(ButtonStyle buttonStyle)
    {
        this.buttonBehavior.setButtonStyle(buttonStyle);
        return this;
    }

    public AjaxButton<T> setButtonSizeModel(IModel<ButtonSize> buttonSizeModel)
    {
        this.buttonBehavior.setButtonSizeModel(buttonSizeModel);
        return this;
    }

    public AjaxButton<T> setButtonStyleModel(IModel<ButtonStyle> buttonStyleModel)
    {
        this.buttonBehavior.setButtonStyleModel(buttonStyleModel);
        return this;
    }
}

File: bootstrap/src/main/kotlin/net/dontdrinkandroot/wicket/bootstrap/component/button/AjaxButton.kt
package net.dontdrinkandroot.wicket.bootstrap.component.button

import net.dontdrinkandroot.wicket.bootstrap.behavior.ButtonBehavior
import net.dontdrinkandroot.wicket.bootstrap.css.ButtonSize
import net.dontdrinkandroot.wicket.bootstrap.css.ButtonStyle
import net.dontdrinkandroot.wicket.markup.html.link.AjaxLink
import net.dontdrinkandroot.wicket.model.model
import org.apache.wicket.ajax.AjaxRequestTarget
import org.apache.wicket.behavior.Behavior
import org.apache.wicket.model.IModel
import org.apache.wicket.model.Model

/**
 * @param <T> Type of the model object.
 */
abstract class AjaxButton<T>(
    id: String,
    model: IModel<T>? = null,
    behaviors: List<Behavior> = emptyList(),
    bodyModel: IModel<String> = Model(null),
    buttonStyleModel: IModel<ButtonStyle> = ButtonStyle.SECONDARY.model(),
    buttonSizeModel: IModel<ButtonSize> = Model(null)
) : AjaxLink<T>(id, model, behaviors, bodyModel) {

    init {
        behaviors.forEach { add(it) }
        add(ButtonBehavior(buttonStyleModel, buttonSizeModel))
    }
}

fun <T> ajaxButton(
    id: String,
    model: IModel<T>? = null,
    behaviors: List<Behavior> = emptyList(),
    bodyModel: IModel<String> = Model(null),
    buttonStyleModel: IModel<ButtonStyle> = ButtonStyle.SECONDARY.model(),
    buttonSizeModel: IModel<ButtonSize> = Model(null),
    onClickHandler: AjaxButton<T>.(target: AjaxRequestTarget?) -> Any?
) = object : AjaxButton<T>(
    id,
    model,
    behaviors = behaviors,
    bodyModel = bodyModel,
    buttonStyleModel = buttonStyleModel,
    buttonSizeModel = buttonSizeModel
) {
    override fun onClick(target: AjaxRequestTarget?) {
        onClickHandler(target)
    }
}

fun ajaxButton(
    id: String,
    bodyModel: IModel<String> = Model(null),
    buttonStyle: ButtonStyle = ButtonStyle.SECONDARY,
    buttonSize: ButtonSize? = null,
    onClickHandler: AjaxButton<Void>.(target: AjaxRequestTarget?) -> Any?
) = object : AjaxButton<Void>(
    id,
    bodyModel = bodyModel,
    buttonStyleModel = Model(buttonStyle),
    buttonSizeModel = Model(buttonSize)
) {
    override fun onClick(target: AjaxRequestTarget?) {
        onClickHandler(target)
    }
}